kafka/main.go

```diff
@@ -67,82 +82,156 @@ func StartConsumers(providedServer *server.Server, logger *zerolog.Logger) error
 		topics = append(topics, topicMapping.Topic)
 	}
 
-	consumerCount, err := strconv.Atoi(os.Getenv("KAFKA_CONSUMERS_PER_NODE"))
-	if err != nil {
-		logger.Error().Err(err).Msg("Failed to convert KAFKA_CONSUMERS_PER_NODE variable to a usable integer. Defaulting to 1.")
-		consumerCount = 1
-	}
+	reader := newConsumer(topics, adsConsumerGroupV1, logger)
 
-	logger.Trace().Msgf("Spawning %d consumer goroutines", consumerCount)
+	// `kafka-go` exposes messages one at a time through its normal interfaces despite
+	// collecting messages with batching from Kafka. To process these messages in
+	// parallel we use the `FetchMessage` method in a loop to collect a set of messages
+	// for processing. Successes and permanent failures are committed. Temporary
+	// failures are not committed and are retried. Miscategorization of errors can
+	// cause the consumer to become stuck forever, so it's important that permanent
+	// failures are not categorized as temporary.
+	for {
+		var (
+			wg           sync.WaitGroup
+			errorResults = make(chan *utils.ProcessingError)
+		)
+		// Any error that occurs while getting the batch won't be available until
+		// the Close() call.
+		ctx := context.Background()
+		batch, err := batchFromReader(ctx, reader, 20, logger)
+		if err != nil {
+			logger.Error().Err(err).Msg("Batching failed")
+			// This should be an error that needs to communicate if its failure is
+			// temporary or permanent. If temporary we need to handle it and if
+			// permanent we need to commit and move on.
+		}
+		for _, msg := range batch {
+			wg.Add(1)
+			logger.Info().Msgf("Processing message for topic %s at offset %d", msg.Topic, msg.Offset)
```

Comment:
This seems like a lot of logging at info, should we switch to debug?

Reply:
yes